// material-design-react/button/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _libMdlHook = require('../lib/mdl-hook');

var _libMdlHook2 = _interopRequireDefault(_libMdlHook);

var Button = (function (_React$Component) {
  _inherits(Button, _React$Component);

  function Button() {
    _classCallCheck(this, _Button);

    _get(Object.getPrototypeOf(_Button.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(Button, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement('button', this.props);
    }
  }]);

  var _Button = Button;
  Button = (0, _libMdlHook2['default'])({
    displayName: 'MDRButton',
    blockClassName: 'mdl-button mdl-js-button',
    modifiers: [{
      prop: 'ripple',
      className: 'mdl-js-ripple-effect',
      type: _react2['default'].PropTypes.bool
      // default: false
    }, {
      prop: 'raised',
      className: 'mdl-button--raised',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'fab',
      className: 'mdl-button--fab',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'miniFab',
      className: 'mdl-button--mini-fab',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'icon',
      className: 'mdl-button--icon',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'colored',
      className: 'mdl-button--colored',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'accent',
      className: 'mdl-button--accent',
      type: _react2['default'].PropTypes.bool
    }, {
      prop: 'display',
      classNameFn: function classNameFn(val) {
        return 'mdl-button--' + val;
      },
      type: _react2['default'].PropTypes.oneOf(['raised', 'fab', 'fabMini', 'icon'])
    }]
  })(Button) || Button;
  return Button;
})(_react2['default'].Component);

exports['default'] = Button;
module.exports = exports['default'];
